week-03/assessment/solar-farm-starter/src/main/java/learn/solarfarm/domain/SolarPanelService.java

```diff
@@ -57,22 +65,31 @@ public SolarPanelResult update(SolarPanel solarPanel) throws DataAccessException
         }
 
         boolean success = repository.update(solarPanel);
-        if (!success) {
+        if (!success)
             result.addErrorMessage("could not update solar panel id " + solarPanel.getId());
-        }
 
         return result;
     }
 
-    // Add a delete method (possibly deleteById?)
-    public SolarPanelResult deleteById(int solarPanelId) throws DataAccessException
+    public SolarPanelResult deleteByKey(String section, int row, int column) throws DataAccessException
     {
         SolarPanelResult result = new SolarPanelResult();
-        boolean success = repository.deleteById(solarPanelId);
-        if (!success)
+        SolarPanel solarPanel = null;
+        try
         {
-            result.addErrorMessage("could not delete solar panel id " + solarPanelId);
+            solarPanel = findByKey(section, row, column);
+            solarPanel.setId(solarPanel.getId());
         }
+        catch(Exception e)
+        {
+            result.addErrorMessage("\n[Err]\nThere is no panel " + section + "-" + row + "-" + column + ".");
+            return result;
+        }
+        boolean success = repository.deleteByKey(solarPanel);
+        if (!success)
+            result.addErrorMessage("could not delete solar panel because solar panel is null");
+        else
+            result.addErrorMessage("\n[Success]\nPanel " + solarPanel.getKey() + " removed.");
```

Comment:
If the repository couldn't delete the solar panel, it's because it couldn't be found, not because it's null. 
```suggestion
        boolean success = repository.deleteByKey(solarPanel);
        if (!success)
            result.addErrorMessage("Solar panel could not be found");
        else
            result.addErrorMessage("\n[Success]\nPanel " + solarPanel.getKey() + " removed.");
```